File: JavaVersion/AoC2020/src/com/company/Day11.java
package com.company;

import java.util.ArrayList;

public class Day11 extends DefaultClass{

    char[][] grid;

    public Day11()
    {
        super(11,true);
        ResetClock();
        ArrayList<String> lines = fr.ReadLine();
        grid = new char[lines.size()][lines.get(0).length()];
        for (int i = 0; i < lines.size(); i++) {
            for (int j = 0; j < lines.get(0).length(); j++) {
                if(lines.get(i).charAt(j)=='L')
                    grid[i][j] = '#';
                else
                    grid[i][j]='.';
            }
        }

        while(simulation());
        int sum=0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if(grid[i][j]=='#')
                    sum++;
            }
        }
        System.out.println("Part1: "+sum);
        TimeResetClock();

        for (int i = 0; i < lines.size(); i++) {
            for (int j = 0; j < lines.get(0).length(); j++) {
                if(lines.get(i).charAt(j)=='L')
                    grid[i][j] = '#';
                else
                    grid[i][j]='.';
            }
        }

        while(simulationpart2());
        sum=0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if(grid[i][j]=='#')
                    sum++;
            }
        }
        System.out.println("Part2: "+sum);
        TimeResetClock();
    }

    private boolean simulationpart2()
    {
        boolean Change = false;
        char[][] grid2 = new char[grid.length][grid[0].length];
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                grid2[i][j]=grid[i][j];
            }
        }
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if(grid[i][j]=='L')
                {
                    if(ChangeLPart2(j,i))
                    {
                        Change=true;
                        grid2[i][j]='#';
                    }
                }
                else if(grid[i][j]=='#')
                {
                    if(ChangehashPart2(j,i))
                    {
                        Change=true;
                        grid2[i][j]='L';
                    }
                }
            }
        }
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                grid[i][j]=grid2[i][j];
            }
        }
        return Change;
    }

    private boolean ChangeLPart2(int x, int y)
    {
        for (int i = x-1; i > -1; i--) {
            if(grid[y][i]=='#')
                return false;
            else if(grid[y][i]=='L')
                i=-1;
        }
        for (int i = x+1; i < grid[0].length; i++) {
            if(grid[y][i]=='#')
                return false;
            else if(grid[y][i]=='L')
                i=grid[0].length;
        }
        for (int i = y-1; i > -1; i--) {
            if(grid[i][x]=='#')
                return false;
            else if(grid[i][x]=='L')
                i=-1;
        }
        for (int i = y+1; i < grid.length; i++) {
            if(grid[i][x]=='#')
                return false;
            else if(grid[i][x]=='L')
                i=grid.length;
        }
        boolean reached = false;
        int x1=x-1,y1=y-1;
        while(!reached)
        {
            try
            {
                if(grid[y1][x1]=='#')
                    return false;
                else if(grid[y1][x1]=='L') {
                    reached = true;
                }
                x1-=1;
                y1-=1;
            }
            catch (Exception E) {
                reached = true;
            }
        }
        reached = false;
        x1=x-1;
        y1=y+1;
        while(!reached)
        {
            try
            {
                if(grid[y1][x1]=='#')
                    return false;
                else if(grid[y1][x1]=='L') {
                    reached = true;
                }
                x1-=1;
                y1+=1;
            }
            catch (Exception E) {
                reached = true;
            }
        }
        reached = false;
        x1=x+1;
        y1=y-1;
        while(!reached)
        {
            try
            {
                if(grid[y1][x1]=='#')
                    return false;
                else if(grid[y1][x1]=='L') {
                    reached = true;
                }
                x1+=1;
                y1-=1;
            }
            catch (Exception E) {
                reached = true;
            }
        }
        reached = false;
        x1=x+1;
        y1=y+1;
        while(!reached)
        {
            try
            {
                if(grid[y1][x1]=='#')
                    return false;
                else if(grid[y1][x1]=='L') {
                    reached = true;
                }
                x1+=1;
                y1+=1;
            }
            catch (Exception E) {
                reached = true;
            }
        }
        return true;
    }

    private boolean ChangehashPart2(int x, int y)
    {
        int sum =0;
        for (int i = x-1; i > -1; i--) {
            if(grid[y][i]=='#') {
                sum++;
                i=-1;
            }
            else if(grid[y][i]=='L')
                i=-1;
        }
        for (int i = x+1; i < grid[0].length; i++) {
            if(grid[y][i]=='#') {
                sum++;
                i=grid[0].length;
            }
            else if(grid[y][i]=='L')
                i=grid[0].length;
        }
        for (int i = y-1; i > -1; i--) {
            if(grid[i][x]=='#') {
                sum++;
                i=-1;
            }
            else if(grid[i][x]=='L')
                i=-1;
        }
        for (int i = y+1; i < grid.length; i++) {
            if(grid[i][x]=='#') {
                sum++;
                i=grid.length;
            }
            else if(grid[i][x]=='L')
                i=grid.length;
        }
        boolean reached = false;
        int x1=x-1,y1=y-1;
        while(!reached)
        {
            try
            {
                if(grid[y1][x1]=='#') {
                    sum++;
                    reached=true;
                }
                else if(grid[y1][x1]=='L') {
                    reached = true;
                }
                x1-=1;
                y1-=1;
            }
            catch (Exception E) {
                reached = true;
            }
        }
        reached = false;
        x1=x-1;
        y1=y+1;
        while(!reached)
        {
            try
            {
                if(grid[y1][x1]=='#') {
                    sum++;
                    reached=true;
                }
                else if(grid[y1][x1]=='L') {
                    reached = true;
                }
                x1-=1;
                y1+=1;
            }
            catch (Exception E) {
                reached = true;
            }
        }
        reached = false;
        x1=x+1;
        y1=y-1;
        while(!reached)
        {
            try
            {
                if(grid[y1][x1]=='#') {
                    sum++;
                    reached=true;
                }
                else if(grid[y1][x1]=='L') {
                    reached = true;
                }
                x1+=1;
                y1-=1;
            }
            catch (Exception E) {
                reached = true;
            }
        }
        reached = false;
        x1=x+1;
        y1=y+1;
        while(!reached)
        {
            try
            {
                if(grid[y1][x1]=='#') {
                    sum++;
                    reached=true;
                }
                else if(grid[y1][x1]=='L') {
                    reached = true;
                }
                x1+=1;
                y1+=1;
            }
            catch (Exception E) {
                reached = true;
            }
        }
        return sum>4;
    }

    private boolean simulation()
    {
        boolean Change = false;
        char[][] grid2 = new char[grid.length][grid[0].length];
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                grid2[i][j]=grid[i][j];
            }
        }
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if(grid[i][j]=='L')
                {
                    if(ChangeL(j,i))
                    {
                        Change=true;
                        grid2[i][j]='#';
                    }
                }
                else if(grid[i][j]=='#')
                {
                    if(Changehash(j,i))
                    {
                        Change=true;
                        grid2[i][j]='L';
                    }
                }
            }
        }
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                grid[i][j]=grid2[i][j];
            }
        }
        return Change;
    }

    private boolean ChangeL(int x, int y)
    {
        for (int i = -1; i < 2; i++) {
            for (int j = -1; j < 2; j++) {
                try
                {
                    if(grid[y+i][x+j]=='#')
                        return false;
                }
                catch (Exception e)
                {

                }
            }
        }
        return true;
    }

    private boolean Changehash(int x, int y)
    {
        int sum=0;
        for (int i = -1; i < 2; i++) {
            for (int j = -1; j < 2; j++) {
                try
                {
                    if(grid[y+i][x+j]=='#')
                        sum++;
                }
                catch (Exception e)
                {

                }
            }
        }
        return sum>4;
    }

}

File: CPPVersion/AoC2020CPP/AoC2020CPP/Day11.cpp
#include "Day11.h"

Day11::Day11() :DefaultClass(11, true)
{
	Layout = FR.ReadCharGrid();
	for (int i = 0; i < Layout.size(); i++)
	{
		for (int c = 0; c < Layout[0].size(); c++)
		{
			if (Layout[i][c] == 'L')
				Layout[i][c] = '#';
		}
	}
	TimeResetClock();
	Part1();
	TimeResetClock();
	Part2();
	TimeClock();
}

void Day11::Part1()
{
	Grid = Layout;
	while (SimulatePart1());
	std::cout << "Part1: " << NumOccupiedSeats() << "\n";
}

void Day11::Part2()
{
	Grid = Layout;
	while (SimulatePart2());
	std::cout << "Part2: " << NumOccupiedSeats() << "\n";
}

int Day11::NumOccupiedSeats()
{
	int sum = 0;
	for (int i = 0; i < Grid.size(); i++)
	{
		for (int c = 0; c < Grid[0].size(); c++)
		{
			if (Grid[i][c] == '#')
				sum++;
		}
	}
	return sum;
}

bool Day11::SimulatePart1()
{
	std::vector<std::vector<char>> Grid2 = std::vector<std::vector<char>>();
	bool change = false;
	for (int i = 0; i < Grid.size(); i++)
	{
		Grid2.push_back(std::vector<char>());
		for (int c = 0; c < Grid[0].size(); c++)
		{
			if (Grid[i][c] == 'L')
			{
				if (ChangeLPart1(c, i))
				{
					Grid2[i].push_back('#');
					change = true;
				}
				else
					Grid2[i].push_back('L');
			}
			else if (Grid[i][c] == '#')
			{
				if (ChangeHashPart1(c, i))
				{
					Grid2[i].push_back('L');
					change = true;
				}
				else
					Grid2[i].push_back('#');
			}
			else
				Grid2[i].push_back('.');
		}
	}
	Grid = Grid2;
	return change;
}

bool Day11::ChangeLPart1(int x, int y)
{
	for (int i = -1; i < 2; i++)
	{
		for (int c = -1; c < 2; c++)
		{
			if (x + c > -1 && x + c<Grid[0].size() && y + i>-1 && y + i < Grid.size())
			{
				if (Grid[y + i][x + c] == '#')
					return false;
			}
		}
	}
	return true;
}

bool Day11::ChangeHashPart1(int x, int y)
{
	int sum=0;
	for (int i = -1; i < 2; i++)
	{
		for (int c = -1; c < 2; c++)
		{
			if (x + c > -1 && x + c<Grid[0].size() && y + i>-1 && y + i < Grid.size())
			{
				if (Grid[y + i][x + c] == '#')
					sum++;
			}
		}
	}
	return sum > 4;
}

bool Day11::SimulatePart2()
{
	std::vector<std::vector<char>> Grid2 = std::vector<std::vector<char>>();
	bool change = false;
	for (int i = 0; i < Grid.size(); i++)
	{
		Grid2.push_back(std::vector<char>());
		for (int c = 0; c < Grid[0].size(); c++)
		{
			if (Grid[i][c] == 'L')
			{
				if (ChangeLPart2(c, i))
				{
					Grid2[i].push_back('#');
					change = true;
				}
				else
					Grid2[i].push_back('L');
			}
			else if (Grid[i][c] == '#')
			{
				if (ChangeHashPart2(c, i))
				{
					Grid2[i].push_back('L');
					change = true;
				}
				else
					Grid2[i].push_back('#');
			}
			else
				Grid2[i].push_back('.');
		}
	}
	Grid = Grid2;
	return change;
}

bool Day11::ChangeLPart2(int x, int y)
{
	int sum = CheckAxis(x, y);
	sum += CheckDiagonal(x, y);
	return sum == 0;
}

bool Day11::ChangeHashPart2(int x, int y)
{
	int sum = CheckAxis(x, y);
	sum += CheckDiagonal(x, y);
	return sum > 4;
}

int Day11::CheckAxis(int x, int y)
{
	int sum = 0;
	int x1=x, y1=y;
	if (x != 0)
	{
		x1 = x - 1;
		while (x1 > 0 && Grid[y][x1] == '.')
			x1--;
		if (Grid[y][x1] == '#')
			sum++;
	}
	if (x != Grid[0].size() - 1)
	{
		x1 = x + 1;
		while (x1 < Grid[0].size() -1 && Grid[y][x1] == '.')
			x1++;
		if (Grid[y][x1] == '#')
			sum++;
	}
	if (y != 0)
	{
		x1 = x;
		y1 = y - 1;
		while (y1 > 0 && Grid[y1][x] == '.')
			y1--;
		if (Grid[y1][x] == '#')
			sum++;
	}
	if (y != Grid.size() - 1)
	{
		x1 = x;
		y1 = y + 1;
		while (y1 < Grid.size() -1 && Grid[y1][x] == '.')
			y1++;
		if (Grid[y1][x] == '#')
			sum++;
	}
	return sum;
}

int Day11::CheckDiagonal(int x, int y)
{
	int sum = 0;
	int x1 = x, y1 = y;
	if (x != 0 && y != 0)
	{
		x1 = x - 1;
		y1 = y - 1;
		while (x1 > 0 && y1 > 0 && Grid[y1][x1] == '.')
		{
			x1--;
			y1--;
		}
		if (Grid[y1][x1] == '#')
			sum++;
	}
	if (x != Grid[0].size() - 1 && y != 0)
	{
		x1 = x + 1;
		y1 = y - 1;
		while (x1 < Grid[0].size()-1 && y1 > 0 && Grid[y1][x1] == '.')
		{
			x1++;
			y1--;
		}
		if (Grid[y1][x1] == '#')
			sum++;
	}
	if (x != Grid[0].size() - 1 && y != Grid.size() - 1)
	{
		x1 = x + 1;
		y1 = y + 1;
		while (x1 < Grid[0].size()-1 && y1 < Grid.size()-1 && Grid[y1][x1] == '.')
		{
			x1++;
			y1++;
		}
		if (Grid[y1][x1] == '#')
			sum++;
	}
	if (x != 0 && y != Grid.size() - 1)
	{
		x1 = x - 1;
		y1 = y + 1;
		while (x1 > 0 && y1 < Grid.size()-1 && Grid[y1][x1] == '.')
		{
			x1--;
			y1++;
		}
		if (Grid[y1][x1] == '#')
			sum++;
	}
	return sum;
}